py/nlrthumb: Do not mark nlr_push as not returning anything.

aykevl opened the pull request:
By adding `__builtin_unreachable()` at the end of `nlr_push`, we're essentially telling the compiler that this function will never return. When GCC link-time optimisation is in use, this means that any time `nlr_push()` is called (which is often), the compiler thinks this function will never return and thus eliminates all code following the call. It breaks the [nrf port](https://github.com/tralamazza/micropython/tree/master/ports/nrf), which uses -flto to reduce code size. When the port compiles with https://github.com/micropython/micropython/commit/97cc48553828ed091325d0d3922eb4cd6c377221, the code size is reduced by about 10K so it was easy to see there's some invalid code elimination going on.

Note: older GCC versions might complain about a missing return statement, but at least version 5.4.1 (Debian stretch) doesn't. If there are any versions that complain, I would propose inserting a return statement anyway for that GCC version and lower as a workaround.

See also:
https://github.com/micropython/micropython/issues/3484
https://github.com/micropython/micropython/pull/3492
https://github.com/micropython/micropython/commit/97cc48553828ed091325d0d3922eb4cd6c377221 (introducing the issue)

aykevl committed:
py/nlrthumb: Do not mark nlr_push as not returning anything.

By adding __builtin_unreachable() at the end of nlr_push, we're
essentially telling the compiler that this function will never return.
When GCC LTO is in use, this means that any time nlr_push() is called
(which is often), the compiler thinks this function will never return
and thus eliminates all code following the call.

Note: I've added a 'return 0' for older GCC versions like 4.6 which
complain about not returning anything (which doesn't make sense in a
naked function). Newer GCC versions (tested 4.8, 5.4 and some others)
don't complain about this.

aykevl commented:
I see one of the tests fails, but it looks unrelated:

```
CC ../../ports/stm32/pin.c
../../ports/stm32/pin.c: In function 'pin_obj_init_helper':
../../ports/stm32/pin.c:379:32: error: 'GPIO_SPEED_FREQ_HIGH' undeclared (first use in this function)
     GPIO_InitStructure.Speed = GPIO_SPEED_FREQ_HIGH;
                                ^
../../ports/stm32/pin.c:379:32: note: each undeclared identifier is reported only once for each function it appears in
make: *** [build/ports/stm32/pin.o] Error 1
make: Leaving directory `/home/travis/build/micropython/micropython/ports/teensy'
The command "make -C ports/teensy" exited with 2.
```

stinos commented:
Since the comment on __builtin_reachable says what it's there for, and that seems a good reason, I don't think simply removing this is acceptable (not sure, I didn't write it). I'd think the fix is to leave it there, conditionally. Like, should be there by default, but shouldn't be there when -flto is active and gcc version is new enough.

dpgeorge commented:
> I see one of the tests fails, but it looks unrelated:

That should be fixed now.  If you rebase on to the latest then Travis should build OK.

dpgeorge commented:
As you see, it's not easy to get this right :)  Especially to support all versions of gcc (of which older ones definitely seem buggy) and also LTO (which seems a bit fragile with respect to hand-written assembler code).



> I'd think the fix is to leave it there, conditionally. Like, should be there by default, but shouldn't be there when -flto is active and gcc version is new enough.

That's also what I'd suggest.  @aykevl are you able to add such a check for the version (not sure which version...) and/or LTO?

aykevl commented:
I've investigated this a bit more, and updated the PR.

Some older GCC versions indeed complain about not returning anything in a naked function. Version 4.6 and lower does, version 5.4 and up doesn't, so I've added a `return 0` only for GCC version < 5.4. Looking at the generated assembly code, returning 0 results in `mov r0, #0` which is never reached but harmless. Using `__builtin_unreachable()` instead eliminates lots of reachable code.

To further verify that `__builtin_unreachable()` marks the function as noreturn, I've created a small gist: https://gist.github.com/aykevl/ecdb91a6ade834879849dde160fa36fc
It is made for an ARM system, like the Raspberry Pi, and can be compiled using `gcc -o lto-test lto-test.c -O2`. With a commented-out `__builtin_unreachabe()` the program does what you would expect. But when you uncomment `__builtin_unreachable()`, everything following the call to `foo()` is removed, effectively marking it noreturn. This is easy to verify in the generated assembly code:

without `__builtin_unreachable()`:

```asm
00010344 <main>:
   10344:       e92d4010        push    {r4, lr}
   10348:       eb000055        bl      104a4 <foo>
   1034c:       e1a01000        mov     r1, r0
   10350:       e59f0008        ldr     r0, [pc, #8]    ; 10360 <main+0x1c>
   10354:       ebffffeb        bl      10308 <printf@plt>
   10358:       e3a00000        mov     r0, #0
   1035c:       e8bd8010        pop     {r4, pc}
   10360:       0001051c        .word   0x0001051c
```

with `__builtin_unreachable()`:

```asm
00010314 <main>:
   10314:       e92d4010        push    {r4, lr}
   10318:       eb00004f        bl      1045c <foo>
```

So all in all I think `__builtin_unreachable()` should never be used in a case like this. Returning something (like 0) is technically invalid, but for all I've seen it doesn't harm either. Only doing that for older GCC versions makes sure we won't add the unnecessary `mov r0, #0` and we will not somehow cause problems in the future.

aykevl commented:
Note: it increases code size on the test machine (using GCC 4.8). I'll test a few things and update this comment.

dpgeorge commented:
@aykevl thanks for the detailed investigation and testing.  I'm happy with the PR as it stands now.

dpgeorge commented on py/nlrthumb.c:
```diff
@@ -76,9 +76,9 @@ __attribute__((naked)) unsigned int nlr_push(nlr_buf_t *nlr) {
     );
 
-    #if defined(__GNUC__)
+    #if __GNUC__ < 4 || (__GNUC__ == 4 && __GNUC_MINOR__ < 8)
```
I just noticed that this will probably not work well with non-GNUC compilers, because the `__GNUC__ < 4` check will probably succeed if `__GNUC__` is not defined.  So I guess it needs to retain the `defined(__GNUC__) bit at the start of the if.

aykevl replied:
Ah, yes, that check should be there. Fixed it.

dpgeorge commented:
Thanks for updating the PR, and thanks for writing a good commit message!  Merged.
